### ISSUE TITLE
Split webserver and convey packages for maintainability

### ISSUE THREAD

**Issue by riannucci** [smartystreets/goconvey#331]
Per issue #329, it may make sense split the `convey` and `goconvey` packages apart, separated by a well-defined protocol. This way the webserver stuff could evolve and improve (in a go server + javascript way), separately from convey, which would evolve in a pure-go/test-framework way (that also happens to speak the web server protocol).

This would be similar to the splitting out of the assertions repo: https://github.com/smartystreets/assertions

WDYT?

**Comment by riannucci** [comment]
(If we do a decent job with the protocol, maybe some other test frameworks could also be encouraged to add support for the UI, too :). That would be a pretty nice "free" upgrade for folks that are using a different framework but understandably don't want to re-write all of their tests just to use the goconvey UI).

**Comment by wallclockbuilder** [comment]
@mdwhatcott  I've extracted the convey package to http://github.com/wallclockbuilder/convey
Please take a look.
Once you approve, please go ahead and transfer the repository from me to the ownership of smartystreets(I've already added you to the repo for this reason).

**Comment by wallclockbuilder** [comment]
All the critical dependency paths are currently 'github.com/wallclockbuilder/...' for testing purposes.

I will follow up with a PR that reverts these dependency paths back to the appropriate smartystreets packages.

**Comment by mdwhatcott** [comment]
Hmm, I didn't expect anything to happen on this so quickly. I'm not sure that we've even defined, much less arrived at the "well-defined protocol" that @riannucci referred to. Thoughts?

@wallclockbuilder - Are you thinking that the import path for the `convey` package would change from

```
github.com/smartystreets/goconvey/convey
```

to

```
github.com/smartystreets/convey
```
?

If so, we should first carefully consider the implications of that breaking change for everyone that currently imports the `convey` package in their test code. Also, there are rumblings that GoConvey could be moved to it's own neutral github organization in the future. Maybe that's the appropriate time to split the project?

I imagine that the project would end up looking something like this:

Web runner (GUI):

```
github.com/GoConvey/goconvey
```

Convey testing library:

```
github.com/GoConvey/convey
```

To summarize, I don't feel like we are ready to make any major structural changes to the project just yet.

But I hope you are still interested in working toward a better project structure in the coming months. @wallclockbuilder

**Comment by wallclockbuilder** [comment]
@riannucci @mdwhatcott What exactly does it mean when you refer to "well-defined protocol"?

**Comment by riannucci** [comment]
Right now convey prints some implementation defined JSON to stdout and uses a process wide mutex to do it. Additionally it buffers the entire convey suite (and all inner Conveys) until the very end of the suite. It's probably not necessary to define a better protocol right away, but since it's an internal implementation detail right now, changing it is easy. If we split the packages in half then it would be harder to make a backwards incompatible change.

A minimal change I would like to see first would be a protocol version so that the UI can at least display a legible warning or error if there is a client server mismatch. Right now I think it will just do "something" :). Aiming for backwards compatibility is also really though for volunteer-run teams to do too, so if we split the packages, we should probably make it clear that users are expected to update both packages in sync.

Stuff like that is what I meant by protocol.

**Comment by wallclockbuilder** [comment]
Thanks @riannucci 
@mdwhatcott Any thoughts?

**Comment by mdwhatcott** [comment]
Yes, thanks for the clarification @riannucci. 

@wallclockbuilder - My thoughts are that this could be a really great direction to take the project. Moving forward with #344 is probably a good place to start.

**Comment by wallclockbuilder** [comment]
:+1:
